### ISSUE TITLE
feat: Separate /model into its own module

### ISSUE THREAD

**Issue by yurishkuro** [jaegertracing/jaeger#4899]
Jaeger repo has always been a monorepo with a single module / go.mod, because it was not meant to be used as a library. But it is often used as a library, in particular in the OpenTelemetry Collector Contrib, where many modules depend on our data model. This creates an undesired coupling, e.g. all our dependencies get pull as OTEL dependencies, sometimes causing conflicts.

In contrast, OTEL's repos are organized as multiple modules, which allows more selective dependencies.

The ask from OTEL project is to make `/model` as separate module (https://github.com/open-telemetry/opentelemetry-collector/issues/8743#issuecomment-1781923246).

I don't know all that it takes to achieve that. Some trial and error will be needed. OTEL repos may serve as a blueprint.

* [ ] Do we need special handling of the `model` module in the main  `go.mod` for local development?
* [ ] How does this change the build?
* [ ] How does this affect code coverage computations?
* [ ] How does this change the release / versioning process?
* [ ] How will dependency updates be handled? OTEL apparently switched from dependabot to renovatebot because of multi-module setup.
* [ ] What additional maintenance changes are needed? E.g. we have `go 1.20` in `go.mod`, which is checked by `scripts/check-go-version.sh` - are there more places like that?

**Comment by AHB102** [comment]
I am new to open source, can I start with this issue ? I am familiar with Go.

**Comment by yurishkuro** [comment]
You can give it a try. You should start by outlining a proposal how this could be done, what changes are required.

**Comment by yurishkuro** [comment]
https://www.jaegertracing.io/get-in-touch/

**Comment by AHB102** [comment]
@yurishkuro I have a example of how I am thinking of resolving this ![jaeger.drawio__1_.png](https://github.com/jaegertracing/jaeger/assets/111697568/81bd794a-401c-483d-bc2b-fc6fa0203b97)

**Comment by kanha-gupta** [comment]
@yurishkuro Any entry point to begin with ? I am new to jaeger :)

**Comment by yurishkuro** [comment]
> OTEL repos may serve as a blueprint.

that is the entry point (e.g. https://github.com/open-telemetry/opentelemetry-collector)

**Comment by kanha-gupta** [comment]
@yurishkuro Thanks, I will come up with a proposal soon :)

**Comment by yurishkuro** [comment]
- I would recommend a prototype first with a fake module, not starting with /model
- I don't think this solution should require import path updates
- I don't know if any additional build or CI changes would be required (thus start with prototype)
- I don't know how the release process would be affected - do we just tag the whole repo with a new version tag as we do today and it would work?
- do we need dependabot changes to keep shared dependencies in sync across modules?

**Comment by kanha-gupta** [comment]
Hi @yurishkuro This is the prototype I have created https://github.com/kanha-gupta/jaegerModelSeperation
It contains go.mod file, go.sum file, A MakeFile and a possible readme (if needed)

**Comment by yurishkuro** [comment]
@kanha-gupta you moved /model into a separate repo, this is explicitly NOT what we're going after. We want to keep `jaeger` as the mono-repo which hosts multiple go-modules.

**Comment by RipulHandoo** [comment]
Hey @yurishkuro, 

I've created a prototype for [/model](https://github.com/jaegertracing/jaeger/tree/main/model). Check out the [demo](https://github.com/RipulHandoo/jaeger/tree/demo/demo). If it looks good, I'll replace `package demo -> package model`. 

Let me know if I'm on the right track and if there are any issues.

**Comment by yurishkuro** [comment]
I just updated the ticket description. To make it clear - it's not rocket science to add a new `go.mod` file under `/model`, this is not what the ticket is about.

**Comment by RipulHandoo** [comment]
Yes, I understand that the task involves adding a new go.mod file under /model. What I've done so far is create a prototype in the /demo directory, where the code is a replica of some files from /model. This was done to check if it aligns with the requirements outlined in the ticket. It's essentially a preliminary step to ensure i'm on the right track.

**Comment by RipulHandoo** [comment]
Added `go.mod` and `go.sum` files to [/model](https://github.com/RipulHandoo/jaeger/tree/separate-model-module/model) and resolved dependencies using `go mod tidy`.

**Comment by shikharish** [comment]
@yurishkuro What is the status of this issue? Can I work on this?

**Comment by RipulHandoo** [comment]
Hey @shikharish you can work on this issue.

**Comment by yurishkuro** [comment]
Dependencies in otel/collector-contrib
```
$ rg -g '*.go' -g '!*_test.go' -N 'github.com\/jaegertracing' | grep -E '^\w' | sed 's/^.*"\(.*\)"$/\1/g' | grep '^github.com' | sort -u

github.com/jaegertracing/jaeger/cmd/agent/app/configmanager
github.com/jaegertracing/jaeger/cmd/agent/app/configmanager/grpc
github.com/jaegertracing/jaeger/cmd/agent/app/httpserver
github.com/jaegertracing/jaeger/cmd/agent/app/processors
github.com/jaegertracing/jaeger/cmd/agent/app/servers
github.com/jaegertracing/jaeger/cmd/agent/app/servers/thriftudp
github.com/jaegertracing/jaeger/cmd/collector/app/sampling
github.com/jaegertracing/jaeger/cmd/collector/app/sampling/samplingstrategy
github.com/jaegertracing/jaeger/model
github.com/jaegertracing/jaeger/model/converter/thrift/zipkin
github.com/jaegertracing/jaeger/pkg/cache
github.com/jaegertracing/jaeger/pkg/metrics
github.com/jaegertracing/jaeger/plugin/sampling/strategyprovider/static
github.com/jaegertracing/jaeger/plugin/storage/es/spanstore/dbmodel
github.com/jaegertracing/jaeger/proto-gen/api_v2
github.com/jaegertracing/jaeger/thrift-gen/agent
github.com/jaegertracing/jaeger/thrift-gen/baggage
github.com/jaegertracing/jaeger/thrift-gen/jaeger
github.com/jaegertracing/jaeger/thrift-gen/zipkincore
```

**Comment by yurishkuro** [comment]
this has been solved by moving model into `jaeger-idl` https://github.com/jaegertracing/jaeger/issues/6571
